File: 22-Binary-Trees-Miscellaneous/python/02_find_median_from_data_stream.py
import heapq

class MedianFinder:

    def __init__(self):
        self.small_heap = []  # max heap (stored as negative values)
        self.large_heap = []  # min heap

    def addNum(self, num: int) -> None:
        heapq.heappush(self.small_heap, -num)
        if len(self.small_heap) - len(self.large_heap) > 1 or (
            self.large_heap and (-self.small_heap[0]) > self.large_heap[0]
        ):
            heapq.heappush(self.large_heap, -heapq.heappop(self.small_heap))
        if len(self.large_heap) - len(self.small_heap) > 1:
            heapq.heappush(self.small_heap, -heapq.heappop(self.large_heap))

    def findMedian(self) -> float:
        if len(self.small_heap) == len(self.large_heap):
            return (-self.small_heap[0] + self.large_heap[0]) / 2.0
        elif len(self.small_heap) > len(self.large_heap):
            return float(-self.small_heap[0])
        else:
            return float(self.large_heap[0])


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: 22-Binary-Trees-Miscellaneous/cpp/02-find-median-from-data-stream.cpp
#include <queue>
#include <vector>

using namespace std;

class MedianFinder {
private:
    priority_queue<int> smallHeap; // max heap
    priority_queue<int, vector<int>, greater<int>> largeHeap; // min heap

public:
    MedianFinder() {}

    void addNum(int num) {
        smallHeap.push(num);
        if (smallHeap.size() > largeHeap.size() + 1 ||
            (!largeHeap.empty() && smallHeap.top() > largeHeap.top())) {
            largeHeap.push(smallHeap.top());
            smallHeap.pop();
        }
        if (largeHeap.size() > smallHeap.size() + 1) {
            smallHeap.push(largeHeap.top());
            largeHeap.pop();
        }
    }

    double findMedian() {
        if (smallHeap.size() == largeHeap.size()) {
            return (smallHeap.top() + largeHeap.top()) / 2.0;
        } else if (smallHeap.size() > largeHeap.size()) {
            return smallHeap.top();
        } else {
            return largeHeap.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */
